File: backend/app/api/v1/contracts.py
"""合同管理 API。"""
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import PermissionChecker
from app.database import get_db
from app.models.user import User
from app.schemas.contract import (
    ContractCompleteRequest,
    ContractCreate,
    ContractListOut,
    ContractOut,
    ContractSignRequest,
    ContractStatsOut,
    ContractTerminateRequest,
    ContractUpdate,
)
from app.services import contract as contract_service

router = APIRouter(prefix="/contracts", tags=["合同管理"])


@router.get("/stats", response_model=ContractStatsOut, summary="合同统计")
def stats(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractStatsOut:
    return ContractStatsOut(**contract_service.contract_stats(db, current_user))


@router.get("", response_model=ContractListOut, summary="合同列表")
def list_contracts(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
    status: Optional[str] = None,
    keyword: Optional[str] = None,
    customer_id: Optional[int] = None,
    scope: Optional[str] = Query(None, description="mine/all"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
) -> ContractListOut:
    total, items = contract_service.list_contracts(
        db,
        current_user,
        status=status,
        keyword=keyword,
        customer_id=customer_id,
        scope_filter=scope,
        page=page,
        page_size=page_size,
    )
    return ContractListOut(total=total, items=[ContractOut.model_validate(x) for x in items])


@router.post("", response_model=ContractOut, summary="起草合同")
def create_contract(
    payload: ContractCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.create_contract(db, current_user, payload)
    return ContractOut.model_validate(contract)


@router.get("/{contract_id}", response_model=ContractOut, summary="合同详情")
def get_contract(
    contract_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.get_contract(db, current_user, contract_id)
    return ContractOut.model_validate(contract)


@router.patch("/{contract_id}", response_model=ContractOut, summary="编辑草稿")
def update_contract(
    contract_id: int,
    payload: ContractUpdate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.update_contract(db, current_user, contract_id, payload)
    return ContractOut.model_validate(contract)


@router.post("/{contract_id}/submit", response_model=ContractOut, summary="提交审批")
def submit_contract(
    contract_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.submit_approval(db, current_user, contract_id)
    return ContractOut.model_validate(contract)


@router.post("/{contract_id}/withdraw", response_model=ContractOut, summary="撤回审批")
def withdraw_contract(
    contract_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.withdraw_approval(db, current_user, contract_id)
    return ContractOut.model_validate(contract)


@router.post("/{contract_id}/approve", response_model=ContractOut, summary="审批通过")
def approve_contract(
    contract_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.approve_contract(db, current_user, contract_id)
    return ContractOut.model_validate(contract)


@router.post("/{contract_id}/reject", response_model=ContractOut, summary="审批驳回")
def reject_contract(
    contract_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
    reason: Optional[str] = None,
) -> ContractOut:
    contract = contract_service.reject_contract(db, current_user, contract_id, reason)
    return ContractOut.model_validate(contract)


@router.post("/{contract_id}/sign", response_model=ContractOut, summary="签署合同")
def sign_contract(
    contract_id: int,
    payload: ContractSignRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.sign_contract(db, current_user, contract_id, payload)
    return ContractOut.model_validate(contract)


@router.post("/{contract_id}/activate", response_model=ContractOut, summary="进入执行")
def activate_contract(
    contract_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.activate_contract(db, current_user, contract_id)
    return ContractOut.model_validate(contract)


@router.post("/{contract_id}/complete", response_model=ContractOut, summary="完成合同")
def complete_contract(
    contract_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
    payload: ContractCompleteRequest = ContractCompleteRequest(),
) -> ContractOut:
    contract = contract_service.complete_contract(db, current_user, contract_id, payload)
    return ContractOut.model_validate(contract)


@router.post("/{contract_id}/terminate", response_model=ContractOut, summary="终止合同")
def terminate_contract(
    contract_id: int,
    payload: ContractTerminateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["contract:view"]))],
) -> ContractOut:
    contract = contract_service.terminate_contract(db, current_user, contract_id, payload)
    return ContractOut.model_validate(contract)

File: frontend/src/api/contracts.ts
/**
 * 合同管理 API
 */
import request from './request'

export type ContractStatus =
  | 'draft'
  | 'pending_approval'
  | 'approved'
  | 'signed'
  | 'active'
  | 'completed'
  | 'terminated'

export interface Contract {
  id: number
  contract_no: string
  title: string
  customer_id: number
  contract_type: string
  amount: number | string
  currency: string
  payment_method?: string | null
  status: ContractStatus | string
  signed_date?: string | null
  effective_date?: string | null
  expire_date?: string | null
  owner_id?: number | null
  creator_id?: number | null
  department_id?: number | null
  approved_by?: number | null
  approved_at?: string | null
  terminate_reason?: string | null
  remark?: string | null
  proof_filename?: string | null
  proof_path?: string | null
  proof_url?: string | null
  created_at: string
  updated_at: string
  customer_name?: string | null
  owner_name?: string | null
  creator_name?: string | null
  approved_by_name?: string | null
  paid_amount?: number | string | null
  next_due_date?: string | null
  collection_status?: string | null
}

export interface ContractStats {
  total: number
  draft: number
  pending_approval: number
  approved: number
  signed: number
  active: number
  completed: number
  terminated: number
  mine: number
}

export const CONTRACT_STATUS_LABEL: Record<string, string> = {
  draft: '草稿',
  pending_approval: '待审批',
  approved: '已审批',
  signed: '已签署',
  active: '执行中',
  completed: '已完成',
  terminated: '已终止',
}

export const CONTRACT_TYPE_OPTIONS = [
  { value: 'ai_product', label: 'AI产品销售' },
  { value: 'ai_custom', label: 'AI定制开发' },
  { value: 'media_ops', label: '自媒体代运营' },
  { value: 'other', label: '其他' },
]

export const PAYMENT_METHOD_OPTIONS = [
  { value: 'once', label: '一次性' },
  { value: 'installment', label: '分期' },
  { value: 'milestone', label: '按里程碑' },
]

export function fetchContractStats() {
  return request.get<ContractStats>('/contracts/stats')
}

export function fetchContracts(params: {
  status?: string
  keyword?: string
  customer_id?: number
  scope?: string
  page?: number
  page_size?: number
}) {
  return request.get<{ total: number; items: Contract[] }>('/contracts', { params })
}

export function fetchContractDetail(id: number) {
  return request.get<Contract>(`/contracts/${id}`)
}

export function createContract(data: Partial<Contract> & { customer_id: number; title: string }) {
  return request.post<Contract>('/contracts', data)
}

export function updateContract(id: number, data: Partial<Contract>) {
  return request.patch<Contract>(`/contracts/${id}`, data)
}

export function submitContract(id: number) {
  return request.post<Contract>(`/contracts/${id}/submit`)
}

export function withdrawContract(id: number) {
  return request.post<Contract>(`/contracts/${id}/withdraw`)
}

export function approveContract(id: number) {
  return request.post<Contract>(`/contracts/${id}/approve`)
}

export function rejectContract(id: number, reason?: string) {
  return request.post<Contract>(`/contracts/${id}/reject`, null, { params: { reason } })
}

export function signContract(
  id: number,
  data?: { signed_date?: string; effective_date?: string; expire_date?: string },
) {
  return request.post<Contract>(`/contracts/${id}/sign`, data || {})
}

export function activateContract(id: number) {
  return request.post<Contract>(`/contracts/${id}/activate`)
}

export function completeContract(id: number) {
  return request.post<Contract>(`/contracts/${id}/complete`, {})
}

export function terminateContract(id: number, reason: string) {
  return request.post<Contract>(`/contracts/${id}/terminate`, { reason })
}
